src/policy/mallocspace/global.rs

```diff
@@ -214,45 +249,180 @@ impl<VM: VMBinding> MallocSpace<VM> {
+    fn map_meta_space_for_chunk(&self, chunk_start: Address) {
+        // Map the metadata space for chunk
+        map_meta_space_for_chunk(&self.metadata, chunk_start);
+
+        // Update the bounds of the max and min chunk addresses seen -- this is used later in the sweep
+        // Lockless compare-and-swap loops perform better than a locking variant
+        let chunk_usize = chunk_start.as_usize();
+        let mut min = self.chunk_addr_min.load(Ordering::Relaxed);
+        let mut max = self.chunk_addr_max.load(Ordering::Relaxed);
+
+        while chunk_usize < min {
+            match self.chunk_addr_min.compare_exchange_weak(
+                min,
+                chunk_usize,
+                Ordering::AcqRel,
+                Ordering::Relaxed,
+            ) {
+                Ok(_) => break,
+                Err(x) => min = x,
+            }
+        }
+
+        while chunk_usize > max {
+            match self.chunk_addr_max.compare_exchange_weak(
+                max,
+                chunk_usize,
+                Ordering::AcqRel,
+                Ordering::Relaxed,
+            ) {
+                Ok(_) => break,
+                Err(x) => max = x,
+            }
+        }
+    }
+
+    pub fn sweep_chunk(&self, chunk_start: Address) {
+        // Call the relevant sweep function depending on the location of the mark bits
+        match VM::VMObjectModel::LOCAL_MARK_BIT_SPEC {
+            MetadataSpec::OnSide(local_mark_bit_side_spec) => {
+                unsafe { self.sweep_chunk_mark_on_side(chunk_start, local_mark_bit_side_spec) };
+            }
+            _ => {
+                unsafe { self.sweep_chunk_mark_in_header(chunk_start) };
+            }
+        }
+    }
 
-        // To sum up the total size of live objects. We check this against the active_bytes we maintain.
+    /// This function is called when the mark bits sit on the side metadata.
+    /// This has been optimized with the use of bulk loading and bulk zeroing of
+    /// metadata.
+    ///
+    /// # Safety
+    /// unsafe as it uses non-atomic accesses to side metadata (although these
+    /// non-atomic accesses should not have race conditions associated with them)
+    /// as well as calling libc functions
+    unsafe fn sweep_chunk_mark_on_side(
+        &self,
+        chunk_start: Address,
+        mark_bit_spec: SideMetadataSpec,
+    ) {
         #[cfg(debug_assertions)]
         let mut live_bytes = 0;
 
-        debug!(
-            "Used bytes before releasing: {}",
-            self.active_bytes.load(Ordering::Relaxed)
-        );
+        debug!("Check active chunk {:?}", chunk_start);
+        let mut chunk_is_empty = true;
+        let mut address = chunk_start;
+        let chunk_end = chunk_start + BYTES_IN_CHUNK;
+        let mut page = conversions::page_align_down(address);
+        let mut page_is_empty = true;
+        let mut on_page_boundary = false;
+        let align: usize = 128 * (1 << LOG_MIN_OBJECT_SIZE);
```

Comment:
Actually you should assert the `log_min_obj_size` for alloc bit spec and mark bit spec are the same, and use the `log_min_obj_size` from one of the specs rather than the constant `LOG_MIN_OBJECT_SIZE`.